csharp/src/Apache.Arrow/Extensions/SpanExtensions.cs

```diff
@@ -14,18 +14,65 @@
 // limitations under the License.
 
 using System;
+using System.Runtime.CompilerServices;
 using System.Runtime.InteropServices;
 
 namespace Apache.Arrow
 {
     public static class SpanExtensions
     {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Span<T> CastTo<T>(this Span<byte> span)
             where T: struct =>
             MemoryMarshal.Cast<byte, T>(span);
 
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static ReadOnlySpan<T> CastTo<T>(this ReadOnlySpan<byte> span)
             where T: struct =>
                 MemoryMarshal.Cast<byte, T>(span);
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this ReadOnlySpan<T> source, Span<T> target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this ReadOnlySpan<T> source, T[] target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this Span<T> source, Span<T> target)
+        {
+            CopyToFix(source, 0, target, 0, source.Length);
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(this Memory<T> source, Memory<T> target)
+        {
+            CopyToFix(source.Span, 0, target.Span, 0, source.Length);
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static void CopyToFix<T>(ReadOnlySpan<T> source, int sourceOffset, Span<T> target, int targetOffset, int length)
+        {
+            for (int i = 0; i < length; ++i)
+            {
+                target[targetOffset + i] = source[sourceOffset + i];
```

Comment:
What happens when `target` and `source` overlap in memory? Does this handle that case?

Reply:
@jkotas suggested pinning the spans before calling `CopyTo`, which should avoid the access violation on .NET Framework. This should avoid issues when the memory overlaps and be more efficient than one-element-at-a-time. Can you try doing that as an alternative fix?

```C#
CopyToFix()
{
    fixed (void* pFrom = MemoryMarshal.GetReference(spanFrom))
    fixed (void* pTo = MemoryMarshal.GetReference(spanTo))
    {
         spanFrom.CopyTo(spanTo);
    }
}
```

Reply:
I'll do some tests - I know the current implementation works.  With out the fix it would crash randomly with in the Arrow library - but almost immediately in the sample code i provided in: dotnet/coreclr#27590

Reply:
> I know the current implementation works

It may work for the cases you tested. But it doesn't work when `target` and `source` overlap in memory. If the beginning of `target` and the end of `source` are the same memory location, by the time the loop gets to the end of `source`, you have overwritten the original values in `source`. Thus the wrong information gets copied to `target`.

Reply:
if they overlap I would expect that to be a program bug (although I can see where that would be useful)  if Span.CopyTo handles overlaps, then yes my simple approach is not correct as a temp copy is needed

Reply:
(this is very clear)
https://docs.microsoft.com/en-us/dotnet/api/system.readonlyspan-1.copyto?view=netcore-3.1

(the remarks should probably match the the above wording)
https://docs.microsoft.com/en-us/dotnet/api/system.span-1.copyto?view=netcore-3.1

I'll test out @jkotas suggestion and change the patch... as that seems like it solves a few problems at once